### ISSUE TITLE
Unavailable volumes returned when they should be filtered out

### ISSUE THREAD

**Issue by codenrhoden** [codedellemc/libstorage#382]
@cduchesne observed this behavior (this is libS 0.4 latest and rexray 0.7 latest):

```
[root@libstorage-rbd-test-admin ~]# ~vagrant/rexray volume ls
ID        Name  Status       Size
rbd.test  test  unavailable  1
[root@libstorage-rbd-test-admin ~]# rbd showmapped
[root@libstorage-rbd-test-admin ~]# ~vagrant/rexray volume attach test
FATA[0001] no volumes found
[root@libstorage-rbd-test-admin ~]# ~vagrant/rexray volume mount test
ERRO[0002] volume in wrong state for attach              host=unix:///var/run/libstorage/147733230.sock instanceID=rbd=172.21.13.10 route=volumeAttach server=sheer-crystal-mq service=rbd storageDriver=rbd task=0 time=1482936753551 tls=false txCR=1482936753 txID=58ab6e6f-f805-41df-5d9b-a1b10963e154
ERRO[0002] volume in wrong state for attach              host=unix:///var/run/libstorage/147733230.sock route=volumeAttach server=sheer-crystal-mq time=1482936753551 tls=false txCR=1482936753 txID=58ab6e6f-f805-41df-5d9b-a1b10963e154
ID  Name  Status  Size  Path
FATA[0002] error mounting volume                         error.status=500 volume=test
[root@libstorage-rbd-test-admin ~]#
```

You can see that the docker integration driver is calling the driver's `attach` method, even though the volume is unavailable.  The driver makes this fail anyways, as of #374, but the call shouldn't be happening in the first place.  The integration driver is looking for volumes with the mask `types.VolAttReqWithDevMapOnlyVolsAttachedToInstanceOrUnattachedVols`.

Hitting the driver manually shows this behavior:

```
# curl -H "Libstorage-Instanceid: rbd=10.50.23.101" localhost:7979/volumes?attachments=31
{
  "rbd": {
    "rbd.redis01": {
      "attachmentState": 3,
      "name": "redis01",
      "size": 8,
      "id": "rbd.redis01",
      "type": "rbd"
    },
    "rbd.vol01": {
      "attachmentState": 4,
      "name": "vol01",
      "size": 5,
      "id": "rbd.vol01",
      "type": "rbd"
    }
  }
}
```

With mask `31`, which is `types.VolAttReqWithDevMapOnlyVolsAttachedToInstanceOrUnattachedVols`, no volumes in `attachmentState` 4 should be returned.  So that's the bug.

Why is this happening? It's a bug in the RBD driver. The libS framework handles processing of attachmentState filtering for most drivers, but RBD sets it's own at https://github.com/codedellemc/libstorage/blob/release/0.4.0/drivers/storage/rbd/storage/rbd_storage.go#L388-L423, which causes the framework to skip it's own filtering (by design).

I need to revisit if having RBD set it's own attachment States is necessary, and if it is, write tests that walk through all the valid scenarios to make sure things like this don't keep happening.

Note that since #374 was merged, nothing bad happens, you just get an uglier error. So it becomes a decision of whether this holds up the libS 0.4.0 release train.

**Comment by akutz** [comment]
HAHAHA. I laugh because of all the work to fix the libStorage framework's filtering and logic to set the attachment state when ultimately the RBD driver bypasses it anyway. *sigh* :)

**Comment by cduchesne** [comment]
I checked further with ScaleIO, and the the problem also exists. The ScaleIO driver blocks in the libstorage attach command, whereas the Ceph driver had no guard rails present. I believe libstorage is supposed to prune the volumes that are not available or attached locally.

rexray cli output:
```
rexray volume ls
ID                Name      Status       Size
925e1ea800000000  vol01     attached     8
925e1ea900000001  vol02     available    8
925e1eaa00000002  vol03     unavailable  8
```

libstorage check with option 31 (VolAttReqWithDevMapOnlyVolsAttachedToInstanceOrUnattachedVols):
```
curl -H "Libstorage-Localdevices: scaleio=" -H "Libstorage-Instanceid: scaleio=c00e3a6100000007" localhost:7979/volumes?attachments=31
{
  "scaleio": {
    "925e1ea800000000": {
      "attachments": [
        {
          "deviceName": "",
          "instanceID": {
            "id": "c00e3a6100000007",
            "driver": "scaleio"
          },
          "status": "",
          "volumeID": "925e1ea800000000"
        }
      ],
      "attachmentState": 2,
      "availabilityZone": "default",
      "name": "vol01",
      "size": 8,
      "id": "925e1ea800000000",
      "type": "default"
    },
    "925e1ea900000001": {
      "attachmentState": 3,
      "availabilityZone": "default",
      "name": "vol02",
      "size": 8,
      "id": "925e1ea900000001",
      "type": "default"
    },
    "925e1eaa00000002": {
      "attachments": [
        {
          "deviceName": "",
          "instanceID": {
            "id": "c00e3a6200000008",
            "driver": "scaleio"
          },
          "status": "",
          "volumeID": "925e1eaa00000002"
        }
      ],
      "attachmentState": 4,
      "availabilityZone": "default",
      "name": "vol03",
      "size": 8,
      "id": "925e1eaa00000002",
      "type": "default"
    }
  }
}
```

**Comment by codenrhoden** [comment]
Hmm, okay I'm happy to dig into this.  If it's happening with ScaleIO as well, there is something else going on.  I did take a look at the filtering code, and I retract a bit of what I said before. While the RBD driver does set it's `attachmentState` directly, this only means that the framework will not override it by looking at the `attachments` field. It is still, however, up to the framework to filter out what volumes get returned to the client based on the mask.

Relevant code that skips `attachmentState` calculation when already set by driver is https://github.com/codedellemc/libstorage/blob/master/api/server/router/volume/volume_routes.go#L179-L207

**Comment by akutz** [comment]
Hi @cduchesne,

To be clear, you're saying that the following volume should not be present, correct?

```json
 "925e1eaa00000002": {
      "attachments": [
        {
          "deviceName": "",
          "instanceID": {
            "id": "c00e3a6200000008",
            "driver": "scaleio"
          },
          "status": "",
          "volumeID": "925e1eaa00000002"
        }
      ],
      "attachmentState": 4,
      "availabilityZone": "default",
      "name": "vol03",
      "size": 8,
      "id": "925e1eaa00000002",
      "type": "default"
}
```

**Comment by akutz** [comment]
Hi @cduchesne,

Are you positive you're using a build with the fix we discussed over the break? Because I don't want to dig into this to find it is the same issue :) This seems shockingly familiar.

**Comment by akutz** [comment]
Hi @cduchesne,

I need you or @codenrhoden to repro the issue and paste a link to the gist with debug logging enabled for the request that should be omitting unavai volumes. I need to see why [this](https://github.com/codedellemc/libstorage/blob/master/api/server/router/volume/volume_routes.go#L165-L174) code-block is not being triggered. Thanks.

In the future please remember to submit debug logs as gists with all potential issues. It helps with troubleshooting, and I will almost always ask for them.

**Comment by akutz** [comment]
Hi @codenrhoden,

You know, it should be fairly simple for us to add a series of unit tests for the [`handleVolumeAttachments`](https://github.com/codedellemc/libstorage/blob/master/api/server/router/volume/volume_routes.go#L131-L177) function in the `github.com/codedellemc/libstorage/blob/master/api/server/router/volume` package. 

I already have several tests for this in the VFS [test package](https://github.com/codedellemc/libstorage/blob/master/drivers/storage/vfs/tests/vfs_test.go#L134-L340), but there appears to be some gap for which I've not accounted.

**Comment by akutz** [comment]
Hi @codenrhoden / @cduchesne,

I think I found the test gap. The test [`TestVolumesWithAttachmentsAttachedAndMineWithNotMyInstanceID `](https://github.com/codedellemc/libstorage/blob/master/drivers/storage/vfs/tests/vfs_test.go#L306-L328) should be asserting this situation, but the test isn't using the mask `31` (`VolAttReqWithDevMapOnlyVolsAttachedToInstanceOrUnattachedVols`), it's using `11` (`VolAttReqOnlyVolsAttachedToInstance `).

When I update the test to use a mask of `31` it fails:

```go
func TestVolumesWithAttachmentsAttachedAndMineWithNotMyInstanceID(
	t *testing.T) {
	tc, _, _, _ := newTestConfigAll(t)
	tf := func(config gofig.Config, client types.Client, t *testing.T) {

		ctx := context.Background()
		iidm := types.InstanceIDMap{
			"vfs": &types.InstanceID{ID: "none", Driver: "vfs"},
		}
		ctx = ctx.WithValue(context.AllInstanceIDsKey, iidm)

		reply, err := client.API().Volumes(ctx, 31)
		//types.VolAttReqOnlyVolsAttachedToInstance)
		if err != nil {
			t.Fatal(err)
		}

		if !assert.Nil(t, reply["vfs"]["vfs-000"]) {
			t.FailNow()
		}
		if !assert.Nil(t, reply["vfs"]["vfs-001"]) {
			t.FailNow()
		}
		if !assert.Nil(t, reply["vfs"]["vfs-002"]) {
			t.FailNow()
		}
	}
	apitests.RunWithClientType(t, types.ControllerClient, vfs.Name, tc, tf)
} 
```

```bash
$ LIBSTORAGE_LOGGING_LEVEL=debug LIBSTORAGE_DEBUG=true go test -tags "gofig pflag libstorage_integration_driver_docker libstorage_storage_driver libstorage_storage_driver_vfs libstorage_storage_executor libstorage_storage_executor_vfs" ./drivers/storage/vfs/tests -run TestVolumesWithAttachmentsAttachedAndMineWithNotMyInstanceID -v

...
--- FAIL: TestVolumesWithAttachmentsAttachedAndMineWithNotMyInstanceID (1.08s)
	vfs_test.go:944: created temp vfs root dir: /var/folders/48/52fjq9r10zx3gnm7j57th0500000gn/T/264891708
	vfs_test.go:957: created temp vfs vol dir: /var/folders/48/52fjq9r10zx3gnm7j57th0500000gn/T/264891708/vol
	vfs_test.go:963: created temp vfs snap dir: /var/folders/48/52fjq9r10zx3gnm7j57th0500000gn/T/264891708/snap
	vfs_test.go:972: created temp vfs dev file: /var/folders/48/52fjq9r10zx3gnm7j57th0500000gn/T/264891708/dev
        Error Trace:    vfs_test.go:323
			tests.go:290
			asm_amd64.s:2184
	Error:  	Expected nil, but got: &types.Volume{Attachments:[]*types.VolumeAttachment(nil), AttachmentState:3, AvailabilityZone:"US", Encrypted:false, IOPS:1000, Name:"Volume 000", NetworkName:"", Size:10240, Status:"", ID:"vfs-000", Type:"myType", Fields:map[string]string{"owner":"root@example.com", "priority":"2"}}
```

The question, @codenrhoden, is why the less restrictive mask works and the more restrictive mask ends up being the more permissive one. Hmmm.

**Comment by codenrhoden** [comment]
Here is a simple paste with the RBD driver, will post a full gist momentarily.

```
DEBU[0002] deferring to driver-specified attachment state  host=unix:///var/run/libstorage/478836241.sock instanceID=rbd=172.21.13.10 route=volumeInspect server=big-ripper-ai service=rbd storageDriver=rbd task=0 time=1482942512934 tls=false txCR=1482942512 txID=15a6c47b-7add-4851-4678-530090a1bc6a
DEBU[0002] including volume                              attachmentState=unavailable host=unix:///var/run/libstorage/478836241.sock instanceID=rbd=172.21.13.10 route=volumeInspect server=big-ripper-ai service=rbd storageDriver=rbd task=0 time=1482942512934 tls=false txCR=1482942512 txID=15a6c47b-7add-4851-4678-530090a1bc6a
DEBU[0002] skipping response schema validation; disabled  host=unix:///var/run/libstorage/478836241.sock instanceID=rbd=172.21.13.10 route=volumeInspect server=big-ripper-ai service=rbd storageDriver=rbd task=0 time=1482942512934 tls=false txCR=1482942512 txID=15a6c47b-7add-4851-4678-530090a1bc6a
DEBU[0002] task completed                                host=unix:///var/run/libstorage/478836241.sock instanceID=rbd=172.21.13.10 route=volumeInspect server=big-ripper-ai service=rbd storageDriver=rbd task=0 time=1482942512934 tls=false txCR=1482942512 txID=15a6c47b-7add-4851-4678-530090a1bc6a
INFO[0002] @ - - [28/Dec/2016:16:28:32 +0000] "GET /volumes/rbd/rbd.test?attachments=31 HTTP/1.1" 200 94
INFO[0002]
INFO[0002]     -------------------------- HTTP REQUEST (SERVER) --------------------------
INFO[0002]     GET /volumes/rbd/rbd.test?attachments=31 HTTP/1.1
INFO[0002]     Host: libstorage-server
INFO[0002]     Accept-Encoding: gzip
INFO[0002]     Libstorage-Instanceid: rbd=172.21.13.10
INFO[0002]     Libstorage-Localdevices: rbd=
INFO[0002]     Libstorage-Tx: txID=15a6c47b-7add-4851-4678-530090a1bc6a, txCR=1482942512
INFO[0002]     User-Agent: Go-http-client/1.1
INFO[0002]
INFO[0002]     -------------------------- HTTP RESPONSE (SERVER) -------------------------
INFO[0002]     Content-Type=application/json
INFO[0002]
INFO[0002]     {
INFO[0002]       "attachmentState": 4,
INFO[0002]       "name": "test",
INFO[0002]       "size": 1,
INFO[0002]       "id": "rbd.test",
INFO[0002]       "type": "rbd"
INFO[0002]     }
```

I wasn't going to try and recreate with SIO, so tried quickly with RBD instead.

**Comment by codenrhoden** [comment]
here's a gist with full output from my run: https://gist.github.com/codenrhoden/b56ca49e8954c4f1effb17378fc6e1b2

**Comment by akutz** [comment]
Thanks. I stopped waiting and handled it myself. Your logs are also appreciated :)

**Comment by akutz** [comment]
Hi @codenrhoden,

I've not implemented a fix, but see the PR #383. It includes a new test that fails when you run the following:

```bash
$ LIBSTORAGE_LOGGING_LEVEL=debug LIBSTORAGE_DEBUG=true go test -tags "gofig pflag libstorage_integration_driver_docker libstorage_storage_driver libstorage_storage_driver_vfs libstorage_storage_executor libstorage_storage_executor_vfs" ./drivers/storage/vfs/tests -v -run TestVolumesWithAttachmentsAttachedAndMineOrUnattachedWithNotMyInstanceID

...

--- FAIL: TestVolumesWithAttachmentsAttachedAndMineOrUnattachedWithNotMyInstanceID (1.07s)
	vfs_test.go:974: created temp vfs root dir: /var/folders/48/52fjq9r10zx3gnm7j57th0500000gn/T/190674608
	vfs_test.go:987: created temp vfs vol dir: /var/folders/48/52fjq9r10zx3gnm7j57th0500000gn/T/190674608/vol
	vfs_test.go:993: created temp vfs snap dir: /var/folders/48/52fjq9r10zx3gnm7j57th0500000gn/T/190674608/snap
	vfs_test.go:1002: created temp vfs dev file: /var/folders/48/52fjq9r10zx3gnm7j57th0500000gn/T/190674608/dev
        Error Trace:    vfs_test.go:353
			tests.go:290
			asm_amd64.s:2184
	Error:  	Expected nil, but got: &types.Volume{Attachments:[]*types.VolumeAttachment(nil), AttachmentState:3, AvailabilityZone:"US", Encrypted:false, IOPS:1000, Name:"Volume 000", NetworkName:"", Size:10240, Status:"", ID:"vfs-000", Type:"myType", Fields:map[string]string{"owner":"root@example.com", "priority":"2"}}
```

Let's figure out why this test is failing and it should fix the issue.

**Comment by cduchesne** [comment]
Here is a gist of same test @codenrhoden ran. Using REX-Ray 0.7.0-rc3

https://gist.github.com/cduchesne/f260119d41a2b23553cb089104978998

```
INFO[0000]     -------------------------- HTTP REQUEST (CLIENT) -------------------------
INFO[0000]     GET /volumes/scaleio/925e1eaa00000002?attachments=31 HTTP/1.1
INFO[0000]     Host: 127.0.0.1:7979
INFO[0000]     Libstorage-Instanceid: scaleio=c00e3a6100000007
INFO[0000]     Libstorage-Localdevices: scaleio=925e1ea800000000::/dev/scinib
INFO[0000]     Libstorage-Tx: txID=385f2e23-91bd-468a-42d1-5a78d3ebc089, txCR=1483553593
INFO[0000]
INFO[0000]
INFO[0000]     -------------------------- HTTP RESPONSE (CLIENT) -------------------------
INFO[0000]     HTTP/1.1 200 OK
INFO[0000]     Content-Length: 350
INFO[0000]     Content-Type: application/json
INFO[0000]     Date: Wed, 04 Jan 2017 18:13:13 GMT
INFO[0000]     Libstorage-Servername: fair-lizard-cx
INFO[0000]
INFO[0000]     {
INFO[0000]       "attachments": [
INFO[0000]         {
INFO[0000]           "deviceName": "",
INFO[0000]           "instanceID": {
INFO[0000]             "id": "c00e3a6200000008",
INFO[0000]             "driver": "scaleio"
INFO[0000]           },
INFO[0000]           "status": "",
INFO[0000]           "volumeID": "925e1eaa00000002"
INFO[0000]         }
INFO[0000]       ],
INFO[0000]       "attachmentState": 4,
INFO[0000]       "availabilityZone": "default",
INFO[0000]       "name": "vol03",
INFO[0000]       "size": 8,
INFO[0000]       "id": "925e1eaa00000002",
INFO[0000]       "type": "default"
INFO[0000]     }
```

**Comment by akutz** [comment]
I know why it's failing to omit that volume. I'm looking at the logic again:

```go
		// if the volume is attached to an instance other than the one provided
		// and only the instance's volumes should be returned, then omit this
		// volume
		if attachments.Mine() &&
			attachments.Attached() &&
			(s == types.VolumeAvailable || s == types.VolumeUnavailable) &&
			!attachments.Unattached() {
			ctx.WithFields(lf).Debug("omitting unavailable volume")
			return false
		}
```

Since the mask includes a bit that says to include unattached vols and the logic only evaluates to omitting the volume IFF the mask does NOT include that bit, the volume is not omitted. I need to adjust that logic branch or introduce a new one. @codenrhoden, what do you think? I've worked on this too long, and I think I no longer have fresh eyes on this issue.

**Comment by akutz** [comment]
Closed with PR #383.
